types/wlr_surface.c

```diff
@@ -0,0 +1,119 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <wayland-server.h>
+#include <wlr/util/log.h>
+#include <wlr/render/interface.h>
+#include <wlr/types/wlr_surface.h>
+
+static void surface_destroy(struct wl_client *client, struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void surface_attach(struct wl_client *client,
+		struct wl_resource *resource,
+		struct wl_resource *buffer, int32_t sx, int32_t sy) {
+	struct wlr_surface *surface = wl_resource_get_user_data(resource);
+	surface->pending_buffer = buffer;
+	surface->pending_attached = true;
+}
+
+static void surface_damage(struct wl_client *client,
+		struct wl_resource *resource,
+		int32_t x, int32_t y, int32_t width, int32_t height) {
+	wlr_log(L_DEBUG, "TODO: surface damage");
+}
+
+static void surface_frame(struct wl_client *client,
+		struct wl_resource *resource, uint32_t callback) {
+	wlr_log(L_DEBUG, "TODO: surface frame");
+}
+
+static void surface_set_opaque_region(struct wl_client *client,
+		struct wl_resource *resource,
+		struct wl_resource *region_resource) {
+	wlr_log(L_DEBUG, "TODO: surface opaque region");
+}
+
+static void surface_set_input_region(struct wl_client *client,
+		struct wl_resource *resource,
+		struct wl_resource *region_resource) {
+
+	wlr_log(L_DEBUG, "TODO: surface input region");
+}
+
+static void surface_commit(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_surface *surface = wl_resource_get_user_data(resource);
+
+	// apply pending state
+	if (surface->pending_attached) {
+		surface->attached = surface->pending_buffer;
+		if (surface->pending_buffer) {
```

Comment:
I would change this a bit. Drop the `pending_attached` variable and just test if `pending_buffer` is not NULL here. I would also rename `attached` to `buffer`.